### DISCUSSION TITLE
シミュレーション速度の向上

### DISCUSSION THREAD

**Discussion by koki0624** [KBKN-Autonomous-Robotics-Lab/orange_ros2#66] [Ideas]
シミュレーションにおける 3D-LiDAR の点群数を減らすことで、シミュレーション速度を向上させることができます。点群数は、xacro ファイルの <gazebo> タグ内で指定して変更することが可能です。

devel/reduct_num_pcd ブランチに、調整したファイルをコミットしました。 2034a1624889c89d67330fa07160d86e95dbc9a0
行なった調整は以下の2点です。
1. 水平方向の点群数を1875から600に削減
2. 垂直16方向のうち上側から7方向を削減

一応、この調整で orange_hosei での slam_toolbox を使った地図作成とナビゲーションは問題なく出来ることを確認しました。
また、私の環境では gazebo 起動時の Real Time Factor（実時間に対するシミュレーション時間の比）が、0.20 から 0.88 程度まで向上しました。

<img width="134" alt="image" src="https://github.com/KBKN-Autonomous-Robotics-Lab/orange_ros2/assets/88423990/3e6ef5fe-e9d0-437b-8a46-7e6337d44cb1">   →   <img width="131" alt="image" src="https://github.com/KBKN-Autonomous-Robotics-Lab/orange_ros2/assets/88423990/2a8d6ea7-3b74-456d-bfb0-0c69d127749e">

ただし、当然実機（Velodyne VLP-16）との仕様に差が生まれるので、場合によっては問題が生じる可能性もあります。

利用するか否かはそれぞれの判断にお任せします。ローカルでの検証時にのみ利用したり、作業ブランチにマージしても問題ないと判断したならマージしても構いません。

なお、noetic でも変更内容はほとんど変わりません。

@KBKN-Autonomous-Robotics-Lab/basic-developers

**Comment by shunki1006** [comment]
実機では本来もっと多くのデータが取れているが必要なものを選択するという考えは良いと思います. 
Nav2のパラメータではmax_beams(障害物処理に利用するセンサデータの点群数(多分))を100にしています. 本日の検証環境でもそのように設定していました. 
100ではいくつになるか検証して頂けないでしょうか.
減らした上での検証が大変でしたら私が月曜日に対応いたしますので教えてください, 上側の削除に関してはこのままでいいと思います.

**Comment by koki0624** [comment]
> 実機では本来もっと多くのデータが取れているが必要なものを選択するという考えは良いと思います. Nav2のパラメータではmax_beams(障害物処理に利用するセンサデータの点群数(多分))を100にしています. 本日の検証環境でもそのように設定していました. 100ではいくつになるか検証して頂けないでしょうか. 減らした上での検証が大変でしたら私が月曜日に対応いたしますので教えてください, 上側の削除に関してはこのままでいいと思います.

amclが用いるのは、3D-LiDARのデータをpointcloud_to_laserscanで2Dのレーザースキャンに変換し、さらにロボット正面の2D-LiDARのデータと合成したレーザースキャンデータです。

現状、pointcloud_to_laserscanのパラメータでは、3D-LiDARの -150° ~ 150° の間を 0.5° でレーザースキャンに変換しています（約300° の間におよそ600点）。これを基準にするなら、3D-LiDARの水平方向の点群数が600（300° の間には500点）だと少し疎なデータになるので、720程度がいいかもしれません。

とはいえ、シミュレーションでは多くの要素が理想的なこともありますし、テストしていく中で自己位置の精度などに問題が生じるようなら点群数を増やしていただければよいかと思います。

**Reply by shunki1006** [reply]
勘違いしていました, ありがとうございます.
